src/gpu/nvidia/cudnn_convolution.cpp

```diff
@@ -91,7 +88,9 @@ status_t cudnn_convolution_fwd_t::execute_convolution(
             args.push_back(with_scratchpad ? sc.memory<void *>(ih, *scratch_acc)
                                            : nullptr);
             args.push_back(pd()->impl_->using_transformed_filter()
-                            ? sc.memory<void *>(ih, *filter_scratch_acc)
+                            ? static_cast<void *>(sc.memory<uint8_t *>(ih,
```

Comment:
From what I understand the issue affects two scenarios:
1. When there are more than 1 scratchpad booked
2. When a user passes sub-buffers as memory arguments

From the SYCL 2020 specification it seems that returning the base pointer is the correct behavior: 
> A sub-buffer object can be created which is a sub-range reference to a base buffer. This sub-buffer can be used to create accessors to the base buffer, which have access to the range specified at time of construction of the sub-buffer.

Also, SYCL doesn't seem to specify any way to extract offset that is used for creating a sub-buffer.

A solution I could came up with is to use ranged accessors. Below is an example of using them in a standalone program:
```cpp
#include <CL/sycl.hpp>

int main() {
    sycl::buffer<std::byte> buf(512);
    auto sub_buf = sycl::buffer<std::byte>(buf, 384, 128);
    auto dev = sycl::device(sycl::gpu_selector{});
    sycl::queue q(dev);

    q.submit([&](sycl::handler &cgh) {
        auto acc_buf_read = buf.get_access<sycl::access::mode::read>(cgh, 512, 0);
        auto acc_sub_buf_read = sub_buf.get_access<sycl::access::mode::read>(cgh, 128, 384);

        std::cout << "acc_buf_read.get_offset(): " << acc_buf_read.get_offset() << std::endl;
        std::cout << "acc_sub_buf_read.get_offset(): " << acc_sub_buf_read.get_offset() << std::endl;

        std::cout << "acc_buf_read.size(): " << acc_buf_read.size() << std::endl;
        std::cout << "acc_sub_buf_read.size(): " << acc_sub_buf_read.size() << std::endl;

        cgh.host_task([=](const sycl::interop_handle &ih) {
            std::byte* buf_ptr = (std::byte*)ih.get_native_mem<sycl::backend::ext_oneapi_cuda>(acc_buf_read);
            std::byte* sub_buf_ptr  = (std::byte*)ih.get_native_mem<sycl::backend::ext_oneapi_cuda>(acc_sub_buf_read);
            std::cout << "buf_ptr: " << buf_ptr << std::endl;
            std::cout << "sub_buf_ptr: " << sub_buf_ptr << std::endl;
        });
    });
    q.wait_and_throw();
    return 0;
}
```
Since memory storage contains the base offset and the buffer the memory storage encapsulates knows its size we can adjust `CTX_OUT_ACCESSOR`, `CTX_IN_ACCESSOR` and `CTX_SCRATCH_ACCESSOR` macros to create the ranged accessors. 
Then we can adjust a function that extracts underlying memory from SYCL accessor via interoperability API:
```cpp
template <typename T, typename U>
    inline T memory(const compat::interop_handle &ih, U acc) {
        auto *ptr = compat::get_native_mem<uint8_t *>(ih, acc) + acc.get_offset();
        return reinterpret_cast<T>(ptr);
}
```

This way we don't have to change any primitives but only a few places to support the two scenarios I mentioned in the beginning.